### ISSUE TITLE
Results on Backtesting

### ISSUE THREAD

**Issue by sf202020sf** [ta4j/ta4j#598]
Hi,

First of all I would like to thank you for all of your effort you spend in this lib - great work.
I have tested the backtesting  - when you run a strategy and analyse tradingRecord i was wondering if you take the right bar for the entry:

 - in the trading records the index of entry is the index where the indicator rule is true (index x)
 - you save the close price(AssetPrice) of the bar x in the trade order

But in general that is not true for live trading:
 - you get bar X in the stream and analyze your indicator and decide to enter the trade
 - then you decide to enter the market and you will get the Open price of bar x+1

Do I get something wrong ? If this is the case then the backtesting results are not correct?

**Comment by nickmitchko** [comment]
I am having the same question. My backtest are slightly off from other platforms regadless of the exact same buy/sell times.

**Comment by Meier-Andre** [comment]
I see it the same as sf202020sf. I developed my own backtesting over the last months with this great lib, and changed it as the way you described it. When my program gets a new tick and recognizes the tick is not in the current bar and is in the next bar, it generates a new bar with index 10 for example in the barseries with my current price and calls the method strategy.shouldEnter(9), because bar nr. 9 is now complete/ended. If the shouldEnter method for bar 9 returns true, my program enters in bar 10 with my current price (open price), of bar 10.

Hopefully this is the right way, what is your opinion?

**Comment by nickmitchko** [comment]
@Meier-Andre That is what would happen in real life correct? Signal after bar at time `t` and then purchase at `t+1`. However, you could also assume that at `t+1` you purchase at the `t` close price. So I'd make sure your using the `t+1` open price.

EG:
```
t-1: shouldEnter() == false --> O: 10 L: 9 H: 11: C: 10
  t: shouldEnter() == true  --> O: 10 L:10 H: 12  C: 12
[You buy here] Buy(price = 12)
t+1: shouldEnter() == false --> O: 12 L: 9 H: 11: C: 10
```

**Comment by Meier-Andre** [comment]
@nickmitchko yes, correct. I buy (in livetrading and in backtesting) at the opening price of bar `t` (EG: 30min Bar from 10:30:00-10:59:59), because my current time (EG: 10:30:01) is already in the new bar `t`  and then i look back if i had a `shouldEnter` signal in bar `t-1` (30min Bar from 10:00:00-10:29:59). 
if `shouldEnter()` in bar `t-1` returns true my stock entry is called in bar `t`,  the following way:
`tradingRecord.enter(series.getEndIndex() [my current bar t 10:30 - 11:00], series.getLastBar().getOpenPrice(),
				100.0, series.getLastBar().getSimpleDateName());`

Thats the way it would work in real live and i developed my backtest the same way, because I have stock data in 2 min bars, in my database and combine them to the lager bars, like 30min bars like in this example, to test different bar sizes.

I take the opening price for trade enter, because the close price is already in the past, when my `tradingRecord.enter` Method is called and if you have day change between bar `t` and `t-1` there can be a larger gap between close of `t-1 `and open of `t`

**Comment by sf202020sf** [comment]
yes I agree and the main problem is also the all statistics are wrong because of this issue - the openprice of the next bar is not equal of the price of the previous bar - I hope somebody can have a look to this issue.

**Comment by nimo23** [comment]
>the main problem is also the all statistics are wrong because of this issue

With statistics, all the criteria-package are ment? 

@team172011 can you confirm this?

**Comment by marsteff** [comment]
I would like to give this thread a new heartbeat.

First aff all, thanks for this very cool framework, i love it.
I played with it a bit and also came to the conclusion, that the backtesting statistics are not that accurate they have to be.

I´ve began with an other approach. On my Bar, i have OHLC Values. I can calculate, if the bar is a green or a red one.
I now have to calculate the actual indicator for bar (for a green bar) with a close price which is the low price, then calculate the close price set to the high price and then calculate the original close price. 
With this approach, i have all values, where the candle is moving in real life (the shouldEnter would be not at the real price like in live trading, but nearly there, but i think this could be calculated too).

The Problem is, that all Indicators using the CachedIndicators, what will give me all the time the same value... Therefore there has to be a possibility to deactivate the caching for backtesting.

I hope my thoughts are clear.

I will create a PR later with this approach.

**Comment by team172011** [comment]
> Therefore there has to be a possibility to deactivate the caching for backtesting.

- When you do backtesting you already have a set of known bars. You cannot travel back in time and change prices - so there should be no need for modifying an existing bar.

> With this approach, i have all values, where the candle is moving in real life

- For real time trading we have the last bar uncached, that means you can update the price of the last bar by using the `BarSeries#addPrice` or `BarSeries#addTrade` methods

**Comment by marsteff** [comment]
@team172011 
I think the point is, that for my understanding, i will emulate the "live-moving" of a candle in the back-testing, because my live-trading logic is also working with moving candles. So when i want to look if my strategy is working fine i also need a backtesting with "moving candles".
Thats why i thought, when i have a list of Bars, i have to simulate the the moving, but i using now a different approach to get the backtesting running.
I am using the same logic like i use in live-trading and feed it with historical data (1min candles in a 15min scale for example). works fine but of course, it is a bit slower.

**Comment by team172011** [comment]
@marsteff thanks for your feedback. This is a frequently asked question. I think we will update the FAQ section regarding this

**Comment by phong-phuong** [comment]
Would the community prefer the BarSeriesManager be updated (along with all the tests that get broken) for the correct logic of opening on the next bar, or would you prefer to have a new class, say. OpenOnNextBarSeriesManager for this?

**Comment by phong-phuong** [comment]
Because a bar is confirmed after it closes. You can only enter on the next
bar. You can't go back in time and open on the the exact close of a closed
bar.
This is precisely why this topic was created.

**Comment by phong-phuong** [comment]
Those that trade will know that markets such as Forex and Stock Markets are closed between sessions and on weekends. So the close of the previous bar does not necessary equal to the open of the next bar.

**Comment by marsteff** [comment]
In our implementation we don't wait for the closing of a candle .. when a rule is satisfied, don't matter at what time, I open a trade and also close a trade, when a rule is satisfied...
When u are trading at day-scale (for example) you could only open/close one trade per day...

I think you miss much money, when you only trade on closing candles..

**Comment by team172011** [comment]
> In our implementation we don't wait for the closing of a candle .. when a rule is satisfied, don't matter at what time, I open a trade and also close a trade, when a rule is satisfied...
When u are trading at day-scale (for example) you could only open/close one trade per day...
I think you miss much money, when you only trade on closing candles..

If you trade on a day-scale the required information is only valid at the end of the day. 
If you want to be able to enter or exit the market during the day, you are not trading on a day-scale. You should change your bar duration to another scale like hour-, minute- or second-scale

**Comment by phong-phuong** [comment]
A more accurate backtest would need to look like this in order for a strategy to have the necessary information to enter a trade:

An incomplete bar at index `t` is presented to the strategy for consideration, this bar has only started trading at the open price, and has yet to move in price. Accessing the high, low, close of the current incomplete bar at `t` will introduces `look-ahead bias`, also known as, seeing into the future.

That way, the strategy has access to the current price which it can execute at (this is typically irrelevant for most people, as the variance is minor most of the time, but very relevant for backtests that need high accuracy)

However, all existing strategies will have to be updated to accomodate this logic, where all bar indexes are shifted by +1, and a strategy that previously required 7 bars for example, would now need at least 8 bars (due to the introduction of the incomplete bar).

Also, unless a more sophisicated backtest is implemented, that prevents users from accessing future prices, users would need to be made aware of any potential look ahead bias when operating on the current bar at index `t`.

**Comment by phong-phuong** [comment]
> In our implementation we don't wait for the closing of a candle .. when a rule is satisfied, don't matter at what time, I open a trade and also close a trade, when a rule is satisfied... When u are trading at day-scale (for example) you could only open/close one trade per day...
> 
> I think you miss much money, when you only trade on closing candles..

The proposed logic changes, shouldn't have an affect on your implementation, as your closing and opening prices on the lower timescale should have little descreprancy.

**Comment by phong-phuong** [comment]
> and not by providing another class `OpenOnNextBarSeriesManager`. We could also add `boolean openOnCurrentBar`-property within `BarSeriesManager` to configure the behaviour explicitly. But I'm afraid that this is also related to the other calculations (indicies, criteria, etc.), so it's probably not enough to just make the changes in the `BarSeriesManager`. We have to take a closer look why this was implemented in that way (= open on the current bar instead of the next).

If the proposed changes were implemented, you would need to carefully examine the impact given your backtest assumes that you can execute on the closing price, but the next tick that you get executed on could be very different price depending on volaility.

**Comment by phong-phuong** [comment]
> > However, it executes trades based on the closing price of bar t, which is **inaccurate** since users are only able to execute trades based on the opening price of t + 1, particularly in markets where gaps frequently occur.
> 
> You're making an assumption here, and in doing so, you're inferring something to justify that assumption. However, if the assumption is already incorrect, then its conclusion is also incorrect. It would be nice if you read and understand my posts here in **this** issue, then you might not have made that assumption.

I'll have to disagree. I have read your posts already and it is you who doesn't understand. When a bar is closed, a new bar is formed when a trade takes place with the last price traded to form the open price (except when trading is closed). Assuming you can get any price you want is the wrong assumption too, unless you use limit orders.

It would be helpful if you learnt a thing or two about how trading actually works before forcing your incorrect assumptions on others. Having access to the next open price is NOT looking into the future, as the bar already opened, and if you were to place a trade, this is most accurate price you can get when working with ohlc bars.  You can only trade at the price presented to you. When the bar opens, you can't go back to trade the price at the close, you have already agreed with me on this.

Ta4j's very simple backtest of presenting competed bars prices at index `t` is the cause of your confusion. In reality, you would be presented with an unfinished bar at index `t`, and the strategies would need to look back at bar `t - 1` for the last completed bar.

However, I didn't want to modify the existing structure, which will require everyone to update their ta4j strategy to accommodate the shifted indexes.

There are a few other inaccuracies and bugs with ta4j I haven't even brought up yet...why would you even choose to add the cost basis as the default implementation for the ReturnCriterion is beyond me. Ta4j doesn't even calculate the return properly when using different sized orders due to developers making wrong assumption and taking short cuts to improve performance.

**Comment by phong-phuong** [comment]
> 

Again you are confused, do re-read my previous post in detail.

**Comment by phong-phuong** [comment]
You took it in the wrong context. I am referring to the next open price for the **trade execution price** for a stratety that trades on the "on open". For "on close" strategies, aka strategies that open before the close, yes that is look ahead.

But have you even developed a live trading strategy? Do you even open the close???

![image](https://github.com/ta4j/ta4j/assets/66398960/498e99a7-668f-47ea-988e-b8df9a63b411)

**Comment by phong-phuong** [comment]
I'm not talking about indicators, I'm talking about where the backtest will execute the price at.

**Comment by phong-phuong** [comment]
I've updated the criteria tests to continue to use the closing prices by setting bar series manager to use the execute on close model.  That way, none of the tests will be affected. I don't have time to update all the tests to pass, nor time to understand how each different developer came up with some of those calculations.

**Comment by phong-phuong** [comment]
I believe you are getting the wrong idea. To clarify, BarSeriesManager now uses a TradeExecutionModel. This model controls whether to execute on the open price of the next candle or close price of the confirmed candle when creating the trading record.

Now a user who uses the default BarSeriesManager will be creating stats off the opening price execution model. This is what you are wanting.

What I'm saying is, the unit tests will continue to use the old `ExecuteOnCloseModel` because I have updated the unit test to use that model by calling BarSeriesManager and passing the `ExecuteOnCloseModel` as a parameter.

Make sense?

**Comment by phong-phuong** [comment]
I see your point, but those criterion tests are testing based on a TradingRecord. If a trading record expects a 10 profit. It has nothing to do with whether BarSeriesManager opened at the open or at the close, and shouldn't even know of it's existence. The tests were designed with closing price in mind and should stay that way.
